peer/connector.go

```diff
@@ -162,6 +230,10 @@ func (conn *connector) allClients(f func(client *upstream.Upstream, e *list.Elem
 
 func (conn *connector) searchClients(f func(client *upstream.Upstream, e *list.Element) bool) {
 	for _, client := range conn.staticClients {
+		if nil == client {
```

Comment:
I think it's reasonable since the client must not be nil. Currently we allocate the slice of clients at the initialization, so if you call this method if will return the nil client although it should not be returned.

Reply:
If you look at the `Initialise` function, you can see that the flow. We allocate a slice for keeping static connections then we try to connect with them to assign value to the allocated slice. So a client in correct way is returned with non-nil value, it should be a real client that you can do something on that without need to check nil before, and no reason for us to get a nil value when we query the clients. It's not expected.

Reply:
Your understanding of how upstream is initiated is correct but this is a function, should this function imply all of what you say ? I do'nt think so. If it imply this, it is coupling. Coupling is not only in the code level , most important it is in abstract level. This is why I say the word in the end , the function should stand alone.